File: src/main/java/com/govtech/csvdata/service/CsvService.java
package com.govtech.csvdata.service;

import com.govtech.csvdata.entity.Invoice;
import com.govtech.csvdata.entity.Task;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVRecord;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

@Slf4j
@Service
public class CsvService {

	private final TaskService taskService;
	private final String dateTimeFormat;

	@Autowired
	public CsvService(@Value("${datetime.format}") String dateTimeFormat, TaskService taskService) {
		this.dateTimeFormat = dateTimeFormat;
		this.taskService = taskService;
	}

	public long saveCsvFile(MultipartFile file) {
		log.info("start saveCsvFile()");
		try (InputStream inputStream = file.getInputStream();
			 BufferedReader fileReader = new BufferedReader(new InputStreamReader(inputStream))) {
			Iterable<CSVRecord> records = CSVFormat.DEFAULT.withFirstRecordAsHeader().parse(fileReader);
			List<Invoice> invoiceList = new ArrayList<>();
			DateTimeFormatter format = DateTimeFormatter.ofPattern(dateTimeFormat);
			log.info("after parsed CSV");
			for (CSVRecord record : records) {
				Invoice invoice = Invoice.builder()
						.invoiceNo(record.get(0))
						.stockCode(record.get(1))
						.description(record.get(2))
						.quantity(Integer.parseInt(record.get(3)))
						.invoiceDate(LocalDateTime.parse(record.get(4), format))
						.unitPrice(new BigDecimal(record.get(5)))
						.customerID(record.get(6))
						.country(record.get(7))
						.build();
				invoiceList.add(invoice);
			}
			log.info("done populating list");
			taskService.saveList(invoiceList);
			int size = invoiceList.size();
			Task task = taskService.createTask(size);
			return task.getId();
		} catch (IOException e) {
			log.error("Encounter error while loading CSV file: {} ", e.getMessage());
			return 0;
		}
	}
}

File: src/main/java/com/govtech/csvdata/service/CsvService.kt
package com.govtech.csvdata.service

import com.govtech.csvdata.entity.Invoice
import org.apache.commons.csv.CSVFormat
import org.apache.commons.csv.CSVRecord
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.math.BigDecimal
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.concurrent.CompletableFuture

@Service
class CsvService(private val taskService: TaskService) {

    val log = LoggerFactory.getLogger(this::class.java)

    @Value("\${datetime.format}")
    private var dateTimeFormat: String? = "M/d/yyyy H:m"

    fun saveCsvFile(file: MultipartFile): Long {
        log.info("start saveCsvFile()")
        try {
            file.inputStream.use { inputStream ->
                BufferedReader(InputStreamReader(inputStream)).use { fileReader ->
                    val records: Iterable<CSVRecord> = CSVFormat.DEFAULT.withFirstRecordAsHeader().parse(fileReader)
                    val invoiceList: MutableList<Invoice> = ArrayList()
                    val format = DateTimeFormatter.ofPattern(dateTimeFormat)
                    log.info("after parsed CSV")
                    for (record in records) {
                        val invoice = Invoice.builder()
                            .invoiceNo(record[0])
                            .stockCode(record[1])
                            .description(record[2])
                            .quantity(record[3].toInt())
                            .invoiceDate(LocalDateTime.parse(record[4], format))
                            .unitPrice(BigDecimal(record[5]))
                            .customerID(record[6])
                            .country(record[7])
                            .build()
                        invoiceList.add(invoice)
                    }
                    log.info("done populating list")
                    CompletableFuture.runAsync { taskService.saveList(invoiceList) }
                    val size = invoiceList.size
                    val task = taskService.createTask(size)
                    return task!!.id
                }
            }
        } catch (e: IOException) {
            log.error("Encounter error while loading CSV file: {} ", e.message)
            return 0
        }
    }
}